### PR TITLE
fix: findDOMNode is deprecated in StrictMode warning when using Button component

### PR DESCRIPTION
**Pull request by MostafaBalata**
**💡 Background and solution

**What is actually happening?**
When using <Button> component a warning message will appear in the browser console "Warning: findDOMNode is deprecated in StrictMode. findDOMNode was passed an instance of Wave which is inside StrictMode. Instead, add a ref directly to the element you want to reference..."

The problem occurs because of Wave component is using a deprecated function `finDOMNode` if you are in StrictMode and using react `^16.13.1`

**What is the solution**
- Remove finDOMNode because of the deprecation.
- Create reference object from React.createRef
- Override the children props by passing the reference

Using <Button> results in "findDOMNode is deprecated in StrictMode" warning

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [x] Perfermance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
There's an open issue here: 
https://github.com/ant-design/ant-design/issues/22493

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     x      |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by MostafaBalata** [comment]
@zombieJ is this PR going to be in consideration soon?

**Comment by yoyo837** [comment]
Maybe not only the buttons.

File: components/_util/wave.tsx [hunk 1]
```diff
@@ -44,8 +47,10 @@ export default class Wave extends React.Component<{ insertExtraNode?: boolean }>
 
   context: ConfigConsumerProps;
 
+  private nodeReference = React.createRef<HTMLElement>();
+
   componentDidMount() {
-    const node = findDOMNode(this) as HTMLElement;
+    const node = this.nodeReference.current as HTMLElement;
```
Comment (zombieJ) [inline components/_util/wave.tsx:53]: This is not equal to `findDOMNode`. Wave children is not always dom node.

Comment (MostafaBalata) [reply components/_util/wave.tsx:53]: but it does the same effect, what is the benefits from findDOMNode, that we can't get after assigning the reference explicitly?

Comment (MostafaBalata) [reply components/_util/wave.tsx:53]: @zombieJ also the tests are passing, which means it wont harm any children that doesn't include dom nodes.
I would like to solve the problem completely, can you guid me with more details?

Comment (zombieJ) [reply components/_util/wave.tsx:53]: > but it does the same effect, what is the benefits from findDOMNode, that we can't get after assigning the reference explicitly?

`findDOMNode` always return dom node but `ref` is not.  It is not same effect.

You can consider to add prop in ConfigProvider to disable wave instead.

**Comment by hengkx** [comment]
#24222

**Comment by HuanxinHu** [comment]
I follow the document https://ant.design/docs/react/use-in-typescript, also have this warning in console, when there is only a button component.

**Comment by hengkx** [comment]
ping

**Comment by zombieJ** [comment]
Close since no more response

**Pull request closed without merging**